javascript/ql/src/semmle/javascript/TypeScript.qll

```diff
@@ -1829,6 +1829,27 @@ class Type extends @type {
     )
   }
 
+  /**
+   * Repeatedly unfolds unions, intersections, and type aliases and gets any of the underlying types,
+   * or this type itself if it is not a union or intersection.
+   *
+   * For example, the type `(S & T) | U` unfolds to `S`, `T`, and `U`.
+   *
+   * If this is a type alias, the alias is itself included in the result, but this is not the case for intermediate type aliases.
```

Comment:
I'm a bit conflicted about including the type alias itself in the result, and not including the union/intersection type in the result.  

Do you have any particular reason for including the type alias itself in the result?

Reply:
One of the really important use-cases is to recognize expressions based on some named type, and the user should be able to specify the name of a type alias here. Like if you want to find all expressions of type `Electron.BrowserWindow` you shouldn't have to worry about whether `BrowserWindow` is a type alias.

So I actually wanted to include all intermediate type aliases, but currently TypeScript only retains the outermost aliased name.

I guess we could rephrase it to say that we do a "best effort" to include all intermediate type aliases. WDYT?

Reply:
> Like if you want to find all expressions of type Electron.BrowserWindow you shouldn't have to worry about whether BrowserWindow is a type alias.

That is a good motivation for why type-aliases are kept, and untion/intersection types are not. :+1: 

> I guess we could rephrase it to say that we do a "best effort" to include all intermediate type aliases. WDYT?

I think I like the current phrasing more.